docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,130 @@
+that by introducing a ``DependencyBear``.
+
+This bear serves as a base class parallelizes tasks for each dependency
+result. A bear dependent on other bears can specify its dependencies in
+``BEAR_DEPS``. For example there are two bears ``Foo`` and ``Bar``
+and bear ``Bar`` depends on ``Foo``. This can be written as
+
+::
+
+    class BarBear(DependencyBear):
+        BEAR_DEPS = {FooBear}
+
+Multiple bears can be included as a dependency of a bear in the BEAR_DEPS
+field. The results of the dependency bears are saved in a dictionary
+called ``_dependency_results`` which is initialized in the ``__init()__``
+method of the class ``Bear.py`` and can be accessed using the method
+``dependency_results()`` also belonging to the same class.
+
+Usage of instance-bound ``Bear.BEAR_DEPS``
+------------------------------------------
+
+A bear might depend on mutilple bears before its execution can begin.
+``Bear.BEAR_DEPS`` is just a collection (a set) of classes (Bear classes) that
+need to be executed before that bear can run. Once all these dependencies
+(Bears) have ran their results are passed into ``self.dependency_results``.
+The results are in the form of a dictionary with the types of the bears and
+their corresponding results (in the form of a list) as *Key-Value* pairs.
+From the previous example if we try to access the BEAR_DEPS of the ``BarBear``
+we will get the following result.
+
+::
+
+    import BarBear
+
+    >>> BarBear.BEAR_DEPS
+    {<class 'coalib.core.Bear.FooBear'>}
+
+In the `__init__()` method of the class ``Bear.py`` the dependencies specified
+in the ``BEAR_DEPS`` are copied to every instance of a Bear run using
+``self.BEAR_DEPS = set(self.BEAR_DEPS)`` which makes runtime modifications
+are possible.
+
+Usage of instance-bound ``Bear.REQUIREMENTS``
+-----------------------------------------
+
+A bear might have some external dependencies that need to be installed before
+the bear can run any analysis. The ``REQUIREMENTS`` field (a set) contains
+instances of the ``PackageRequirement`` class from the coala's
+``dependency_management`` module. For example.
+
+::
+
+    >>> from dependency_management.requirements.PackageRequirement import (
+    ...     PackageRequirement)
+    >>> class BarBear(Bear):
+    ...     REQUIREMENTS = {
+    ...         PackageRequirement('pip', 'coala_decorators', '0.2.1')}
+
+If the bear uses requirements from a package manager then we can use the
+subclass of the ``PackageRequirement`` class for that package manager. For
+example in case of a pip requirement do the following:
+
+::
+
+    >>> from dependency_management.requirements.PipRequirement import (
+    ...     PipRequirement)
+    >>> class FooBear(Bear):
+    ...     REQUIREMENTS = {PipRequirement('coala_decorators', '0.2.1')}
+
+The ``check_prerequisites()`` method in ``Bear.py`` is responsible for
+checking whether all the required packages are installed. If not then
+a runtime exception is raised.
```

Comment:
Just to note: This is still not merged, and it's actually quite problematic, because `check_prerequisites` is called on `bear.__init__`. So this one should be left out for now until I decide what to do (or someone wants to discuss that :smile:)^^